scripts/proxy/proxy.go

```diff
@@ -241,32 +413,126 @@ func configureOAuth(clientId, clientSecret string, scopes []string, providerInfo
 	}
 }
 
+func setupOAuth2ClientCredentialsHttpClient(cfg *clientcredentials.Config, provider string) common.AuthenticatedHTTPClient {
+	ctx := context.Background()
+
+	options := []common.OAuthOption{
+		common.WithOAuthClient(http.DefaultClient),
+		common.WithTokenSource(cfg.TokenSource(ctx)),
+		common.WithOAuthDebug(debugRequest),
+	}
+
+	oauthClient, err := common.NewOAuthHTTPClient(ctx, options...)
+	if err != nil {
+		panic(err)
+	}
+
+	conn, err := connector.NewConnector(
+		provider,
+		connector.WithAuthenticatedClient(oauthClient),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	return conn.HTTPClient().Client
+}
+
 // This helps with refreshing tokens automatically.
-func setupHttpClient(cfg *oauth2.Config, tokens *oauth2.Token, provider string, providerInfo *providers.ProviderInfo) *http.Client {
+func setupOAuth2AuthCodeHttpClient(cfg *oauth2.Config, tokens *oauth2.Token, provider string) common.AuthenticatedHTTPClient {
 	ctx := context.Background()
 
 	conn, err := connector.NewConnector(
 		provider,
-		connector.WithClient(ctx, http.DefaultClient, cfg, tokens),
+		connector.WithClient(ctx, http.DefaultClient, cfg, tokens, common.WithOAuthDebug(debugRequest)),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	return conn.HTTPClient().Client
+}
+
+func basicAuth(username, password string) string {
+	auth := username + ":" + password
+	return base64.StdEncoding.EncodeToString([]byte(auth))
+}
+
+func setupBasicAuthHttpClient(provider string, user, pass string) common.AuthenticatedHTTPClient {
+	client, err := common.NewHeaderAuthHTTPClient(context.Background(),
```

Comment:
Right now, the heavy lifting is done by the user of the connector, not the connector itself. Ideally it should be the other way around. The code here for constructing the headers, doing base64 encoding, etc. now has to be repeated inside the server (and any future codebases that uses the connectors library)

I'm imagining that instead of `common.NewHeaderAuthHTTPClient`, we have:
- `common.NewBasicAuthHttpClient`
- `common.NewApiKeyAuthHttpClient`

Usage would be something like:
```
client, err := common.NewBasicAuthHttpClient(context.Background(), provider, common.WithDebug(common.DefaultDebug)) // see my other comment about a default debug function
```
The implementation of NewBasicAuthHttpClient would include looking at ProviderInfo and constructing the appropriate headers.

Reply:
Got it. I've made it much more centralized in this version. Everyone can just use `ProviderInfo.NewClient` which will handle all of the construction logic. I also implemented your suggestion for dedicated basic and api key clients.